src/mscorlib/src/System/String.Manipulation.cs

```diff
@@ -656,6 +636,11 @@ public unsafe static string Join(string separator, string[] value, int startInde
 
         private unsafe static string JoinCore(char* separator, int separatorLength, object[] values)
         {
+            // If the separator is null, it is converted to an empty string before entering this function.
+            // Even for empty strings, fixed should never return null (it should return a pointer to a null char).
+            Contract.Assert(separator != null);
+            Contract.Assert(separatorLength >= 0);
```

Comment:
Should this be `separatorLength > 0` instead? It looks like after your changes, `Concat` is called if the separator is null or empty.